src/plugins/vis_types/xy/public/expression_functions/x_domain.ts

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { uniq } from 'lodash';
+import { unitOfTime } from 'moment';
+import { ExpressionValueVisDimension } from '../../../../visualizations/common';
+import type { ExpressionFunctionDefinition, Datatable } from '../../../../expressions/public';
+import { getAdjustedInterval } from '../../common';
+import { XDomainArguments, ExpressionValueXDomain } from '../types';
+import { getValueByAccessor } from '../utils/accessors';
+
+export const X_DOMAIN_EXPRESSION = 'x_domain';
+
+export const getAdjustedDomain = (
+  data: Datatable['rows'],
+  column: ExpressionValueVisDimension,
+  domain: {
+    intervalValue: number;
+    intervalUnit: string;
+    min: number;
+    max: number;
+    minInterval?: number;
+  },
+  timezone?: string,
+  considerInterval?: boolean
+) => {
+  const values = uniq(data.map((row) => getValueByAccessor(row, column.accessor)).sort());
+  const [first] = values;
+  const last = values[values.length - 1];
+  const domainMin = Math.min(first, domain.min);
+  const domainMaxValue = Math.max(domain.max - (domain.minInterval ?? 0), last);
+  const domainMax = considerInterval ? domainMaxValue + (domain.minInterval ?? 0) : domainMaxValue;
+  const minInterval = getAdjustedInterval(
+    values,
+    domain.intervalValue,
+    domain.intervalUnit as unitOfTime.Base,
+    timezone ?? ''
+  );
+
+  return {
+    min: domainMin,
+    max: domainMax,
+    minInterval,
+  };
+};
+
+export const xDomain = (): ExpressionFunctionDefinition<
+  'x_domain',
+  Datatable | null,
+  XDomainArguments,
+  ExpressionValueXDomain
+> => ({
+  name: X_DOMAIN_EXPRESSION,
+  help: i18n.translate('visTypeXy.function.valueaxis.help', {
+    defaultMessage: 'Generates value axis object',
+  }),
+  type: X_DOMAIN_EXPRESSION,
+  args: {
+    minInterval: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.minInterval.help', {
+        defaultMessage: 'Min interval',
+      }),
+    },
+
+    intervalUnit: {
+      types: ['string'],
+      help: i18n.translate('visTypeXy.function.xDomain.intervalUnit.help', {
+        defaultMessage: 'intervalUnit',
+      }),
+    },
+
+    intervalValue: {
+      types: ['number'],
+      help: i18n.translate('visTypeXy.function.xDomain.intervalValue.help', {
+        defaultMessage: 'intervalValue',
+      }),
+    },
+
+    column: {
+      types: ['vis_dimension'],
+      help: i18n.translate('visTypeXy.function.column.intervalValue.help', {
+        defaultMessage: 'column',
```

Comment:
Can we describe a little bit better what column is exactly? I am also confused why the `i18n.translate('visTypeXy.function.column.intervalValue.help'` has the intervalValue key?

Reply:
@stratoula, I've added the description to all the fields.